File: Languages/npeg_php/StringInputIterator.php
<?
namespace RobustHaven\Text\Npeg;

class StringInputIterator extends InputIterator
{
	private $input;
	
	public function __construct($input)
	{
		parent::__construct(count($input));
		$this->input = $input;
	}
		public abstract function Text($start, $end);

	public function Text($start, $end)
	{
		$n = $end - $start;
		$bytes = array();
		for($i = 0; $i < $n; $i++)
		{
			$bytes[$i] = $this->input[$start+$i];
		}
		return $bytes;
	}
	
	public function Current()
	{
		 if ($this->index >= $this->length) return -1;
            else 
            {
                return $this->input[$this->index];
            }
	}
	
	public  function Next()
	{
		if ($this->index >= $this->length) return -1;
        else 
        {
        	$this->index += 1;
            if ($this->index >= $this->length) return -1;
            else return $this->input[$this->index];
         }
	}
	
	public  function Previous()
	{
		if ($this->index <= 0)
        {
        	return -1;
        }
        else 
        {
        	if($this->Length == 0)throw new Exception();
            $this->index -= 1;
            return $this->input[$this->index];
         }	
	}
}

?>

File: Languages/npeg_csharp/RobustHaven.Text.Npeg/StringInputIterator.cs
using System;
using System.Diagnostics;
using System.Text;

namespace RobustHaven.Text.Npeg
{
	public class StringInputIterator : InputIterator
	{
		private readonly Byte[] input;

		public StringInputIterator(Byte[] input)
			: base(input.Length)
		{
			this.input = input;
		}

		public StringInputIterator(String us_ascii)
			: base(Encoding.ASCII.GetBytes(us_ascii).Length)
		{
			input = Encoding.ASCII.GetBytes(us_ascii);
			Index = 0;
		}


		public override byte[] Text(int start, int end)
		{
			var bytes = new byte[end - start];
			Array.Copy(input, start, bytes, 0, end - start);
			return bytes;
		}


		public override short Current()
		{
			if (Index >= Length) return -1;
			else
			{
				return input[Index];
			}
		}

		public override short Next()
		{
			if (Index >= Length) return -1;
			else
			{
				Index += 1;

				if (Index >= Length) return -1;
				else return input[Index];
			}
		}

		public override short Previous()
		{
			if (Index <= 0)
			{
				return -1;
			}
			else
			{
				Debug.Assert(Length > 0);

				Index -= 1;

				return input[Index];
			}
		}
	}
}